File: articles/situationism-moral-character-and-the-stability-of-virtue/fortran/moral_action_probability.f90
! Synthetic situationism moral action probability example.
! Educational scaffold only.
!
! Compile:
!   gfortran fortran/moral_action_probability.f90 -o outputs/tables/moral_action_probability_fortran
! Run:
!   ./outputs/tables/moral_action_probability_fortran

program moral_action_probability
  implicit none

  real :: character_strength
  real :: habit_strength
  real :: situational_pressure
  real :: voice_safety
  real :: accountability_visibility
  real :: depletion
  real :: latent
  real :: probability

  character_strength = 0.8
  habit_strength = 0.6
  situational_pressure = 0.7
  voice_safety = 0.4
  accountability_visibility = 0.5
  depletion = 0.3

  latent = 0.42 * character_strength + &
           0.25 * habit_strength - &
           0.55 * situational_pressure + &
           0.25 * voice_safety + &
           0.25 * accountability_visibility - &
           0.20 * depletion - &
           0.22 * character_strength * situational_pressure

  probability = 1.0 / (1.0 + exp(-latent))

  print *, "Synthetic moral action probability:", probability
end program moral_action_probability

File: articles/what-is-moral-psychology/rust/moral_action_probability.rs
// Synthetic moral action probability.
// Educational scaffold only.
//
// Compile:
//   rustc rust/moral_action_probability.rs -o outputs/tables/moral_action_probability_rust
// Run:
//   ./outputs/tables/moral_action_probability_rust

fn logistic(x: f64) -> f64 {
    1.0 / (1.0 + (-x).exp())
}

fn main() {
    let moral_salience = 0.7_f64;
    let moral_judgment = 0.8_f64;
    let moral_emotion = 0.5_f64;
    let moral_identity = 0.6_f64;
    let self_regulation = 0.8_f64;
    let institutional_pressure = 0.4_f64;

    let latent_action =
        0.25 * moral_salience
        + 0.30 * moral_judgment
        + 0.15 * moral_emotion
        + 0.20 * moral_identity
        + 0.20 * self_regulation
        - 0.25 * institutional_pressure;

    let probability = logistic(latent_action);

    println!("Synthetic moral action probability: {:.3}", probability);
}
